File: June2021/C++/Max Area of Island.cpp
/**
 * @author ssrbazpur
 * Time Complexity: O(M*N)
 * Space COmplexity:O(M*N)
 */

class Solution
{
public:
    void dfs(vector<vector<int>> &grid, int row, int col, int &ans)
    {

        if (row < 0 || row > grid.size() - 1 || col < 0 || col > grid[0].size() - 1)
        {
            return;
        }

        if (grid[row][col] == 0)
        {
            return;
        }
        if (grid[row][col] == 1)
        {
            grid[row][col] = 0;
            ans = ans + 1;
        }

        dfs(grid, row + 1, col, ans);
        dfs(grid, row - 1, col, ans);
        dfs(grid, row, col - 1, ans);
        dfs(grid, row, col + 1, ans);
    }
    int maxAreaOfIsland(vector<vector<int>> &grid)
    {

        int ans = 0;
        int result = 0;
        for (int i = 0; i < grid.size(); i++)
        {
            for (int j = 0; j < grid[i].size(); j++)
            {
                ans = 0;
                if (grid[i][j] == 1)
                {

                    dfs(grid, i, j, ans);
                }
                result = max(result, ans);
            }
        }

        return result;
    }
};

File: July2022/Java/MaxAreaOfIsland.java

// @saorav21994
// TC: O(R*C)
// SC: O(R*C)

class Solution {
    
    public int dfs(int [][]grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[i].length || grid[i][j] == 0) {
            return 0;
        }
        grid[i][j] = 0;     // marked as visited
        int ctr = 1;
        ctr += dfs(grid, i+1, j);
        ctr += dfs(grid, i, j+1);
        ctr += dfs(grid, i-1, j);
        ctr += dfs(grid, i, j-1);
        return ctr;
    }
    
    public int maxAreaOfIsland(int[][] grid) {
        int max = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                max = Math.max(max, dfs(grid, i, j));
            }
        }
        return max;
    }
}
